dbms/src/Flash/Mpp/MinTSOScheduler.cpp

```diff
@@ -268,42 +285,49 @@ bool MinTSOScheduler::scheduleImp(
     const MPPQueryId & query_id,
     const MPPGatherTaskSetPtr & query_task_set,
     MPPTaskScheduleEntry & schedule_entry,
-    const bool isWaiting,
+    bool isWaiting,
     bool & has_error)
 {
     auto needed_threads = schedule_entry.getNeededThreads();
+    auto & entry = getOrCreateGroupEntry(query_id.resource_group_name);
+
     auto check_for_new_min_tso
-        = query_id <= min_query_id && estimated_thread_usage + needed_threads <= thread_hard_limit;
+        = query_id <= entry.min_query_id && global_estimated_thread_usage + needed_threads <= thread_hard_limit;
     auto check_for_not_min_tso
-        = (active_set.size() < active_set_soft_limit || active_set.find(query_id) != active_set.end())
-        && (estimated_thread_usage + needed_threads <= thread_soft_limit);
+        = (entry.active_set.size() < active_set_soft_limit || entry.active_set.find(query_id) != entry.active_set.end())
```

Comment:
`active_set_soft_limit` and `thread_soft_limit` is a global limit for all resource groups?

Reply:
`active_set_soft_limit` and `thread_soft_limit` are not global. They are only same value for all resource groups. So I didn't put them into `GroupEntry`

`thread_hard_limit` is global, so every scheduled task will cause adding to `global_estimated_thread_usage`, which is also global for all resource groups

Reply:
OK, got it